tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,708 @@
+	inBufferChunksLocks [inBufferShards]sync.RWMutex
+
+	// Reader.
+	bs map[int]ByteSlice
+	cs map[int]io.Closer // Closers for resources behind the byte slices.
+
+	size  int64 // The total size of bytes in the reader.
+	bsMtx sync.RWMutex
+	pool  chunkenc.Pool
+
+	quit chan struct{}
+}
+
+const (
+	// HeadSegmentHeaderSize is the total size of the header for the segment file.
+	HeadSegmentHeaderSize = SegmentHeaderSize + 16
+	// HeaderMintOffset is the offset where the first byte of MinT for segment file exists.
+	HeaderMintOffset = SegmentHeaderSize
+	// HeaderMaxtOffset is the offset where the first byte of MaxT for segment file exists.
+	HeaderMaxtOffset = HeaderMintOffset + 8
+	// MaxSegmentSize is the max size of a segment file.
+	// Setting size to the max int32 as setting it to max int 64 crashes 64 systems too.
+	MaxSegmentSize = math.MaxInt32
+)
+
+// NewHeadReadWriter returns a new writer against the given directory
+// using the default segment time.
+func NewHeadReadWriter(dir string, pool chunkenc.Pool) (*HeadReadWriter, error) {
+	return newHeadReadWriter(dir, DefaultHeadChunkSegmentTime, pool)
+}
+
+func newHeadReadWriter(dir string, segmentTime int64, pool chunkenc.Pool) (*HeadReadWriter, error) {
+	if segmentTime <= 0 {
+		segmentTime = DefaultHeadChunkSegmentTime
+	}
+
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return nil, err
+	}
+	dirFile, err := fileutil.OpenDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	hrw := &HeadReadWriter{
+		dirFile:     dirFile,
+		n:           0,
+		segmentTime: segmentTime,
+		pool:        pool,
+		quit:        make(chan struct{}),
+	}
+
+	for i := 0; i < inBufferShards; i++ {
+		hrw.inBufferChunks[i] = make(map[uint64]chunkenc.Chunk)
+	}
+
+	if err := hrw.initReader(); err != nil {
+		return nil, err
+	}
+
+	return hrw, nil
+}
+
+func (w *HeadReadWriter) initReader() (err error) {
+	bs := map[int]ByteSlice{}
+	cs := map[int]io.Closer{}
+	defer func() {
+		if err != nil {
+			var merr tsdb_errors.MultiError
+			merr.Add(err)
+			merr.Add(closeAllFromMap(cs))
+			err = merr
+		}
+	}()
+
+	files, err := sequenceFilesMap(w.dirFile.Name())
+	if err != nil {
+		return err
+	}
+	if w.pool == nil {
+		w.pool = chunkenc.NewPool()
+	}
+
+	seqs := make([]int, 0, len(files))
+	for seq, fn := range files {
+		f, err := fileutil.OpenMmapFile(fn)
+		if err != nil {
+			return errors.Wrap(err, "mmap files")
+		}
+		cs[seq] = f
+		bs[seq] = realByteSlice(f.Bytes())
+		seqs = append(seqs, seq)
+	}
+
+	w.bs = bs
+	w.cs = cs
+	w.size = 0
+
+	// Check for unsequential files.
+	sort.Ints(seqs)
+	if len(seqs) == 0 {
+		return nil
+	}
+	lastSeq := seqs[0]
+	for _, seq := range seqs[1:] {
+		if seq != lastSeq+1 {
+			return errors.Errorf("found unsequential segment files %d and %d", lastSeq, seq)
+		}
+		lastSeq = seq
+	}
+
+	for i, b := range w.bs {
+		if b.Len() < HeadSegmentHeaderSize {
+			return errors.Wrapf(errInvalidSize, "invalid segment header in segment %d", i)
```

Comment:
Note to self: have repair for these errors too after https://github.com/prometheus/prometheus/pull/6830 is merged